File: adapters/claude-code/hooks.js
#!/usr/bin/env node
/**
 * Yeelight Vibe Bridge — Claude Code adapter (Node.js)
 */
const fs = require("fs");
const path = require("path");
const os = require("os");
const http = require("http");
const BRIDGE = "http://127.0.0.1:9877";
const LOG = path.join(os.homedir(), ".yeelight-vibe-bridge", "hook-debug.log");
const READ_TOOLS = new Set(["Read","LS","Grep","Glob","Task","TodoRead","NotebookRead"]);
const WRITE_TOOLS = new Set(["Write","Edit","NotebookEdit"]);
const EXEC_TOOLS = new Set(["Bash","PowerShell"]);  // command execution tools
const WEB_TOOLS = new Set(["WebSearch","WebFetch"]);
const WEB_KW = ["curl","wget","http ","fetch","npx ","npm ","pip "];

function log(msg) {
  try { fs.appendFileSync(LOG, new Date().toISOString() + " " + msg + "\n"); } catch(_) {}
}

function post(path, data) {
  log("POST " + path + " " + JSON.stringify(data));
  try {
    http.request(BRIDGE + path, {
      method: "POST", headers: { "Content-Type": "application/json" }, timeout: 300
    }).end(JSON.stringify(data));
  } catch (_) {}
}

function readStdin() {
  return new Promise((resolve) => {
    let resolved = false, timer = null;
    const done = (v) => {
      if (resolved) return;
      resolved = true;
      if (timer) clearTimeout(timer);
      resolve(v);
    };
    let data = "";
    process.stdin.on("readable", () => {
      let chunk;
      while ((chunk = process.stdin.read()) !== null) data += chunk;
      try { done(JSON.parse(data.trim())); return; } catch (_) {}
    });
    process.stdin.on("end", () => {
      try { done(JSON.parse(data.trim())); } catch (_) { done(null); }
    });
    process.stdin.on("error", () => done(null));
    timer = setTimeout(() => { try { done(JSON.parse(data.trim())); } catch (_) { done(null); } }, 1000);
    process.stdin.resume();
  });
}

function toolState(name, input) {
  if (READ_TOOLS.has(name)) return "reading";
  if (WRITE_TOOLS.has(name)) return "writing";
  if (EXEC_TOOLS.has(name)) {
    const cmd = (typeof input?.command === "string" ? input.command : "").toLowerCase();
    return WEB_KW.some(kw => cmd.startsWith(kw)) ? "fetching" : "executing";
  }
  if (WEB_TOOLS.has(name)) return "fetching";
  return "thinking";
}

async function main() {
  const mode = process.argv[2]?.toLowerCase();
  if (!mode) return;
  const base = { pid: "claude-hook" };

  log("=== " + mode + " ===");

  switch (mode) {
    case "user_prompt":
      post("/api/state", { ...base, state: "thinking" });
      break;
    case "pre_tool": {
      const ev = await readStdin();
      if (!ev) { log("  ev=null → thinking"); post("/api/state", { ...base, state: "thinking" }); break; }
      log("  ev=" + JSON.stringify(ev));
      // Check ALL possible permission-related fields
      const perm = ev.permissionDecision || ev.permission_decision
                || ev.permission_required || ev.requires_permission
                || ev.permission || ev.decision || "";
      log("  perm=" + (perm || "(none)"));
      if (perm === "ask") { log("  → waiting"); post("/api/state", { ...base, state: "waiting" }); break; }
      const tn = ev.tool_name || ev.toolName || "";
      const ti = ev.tool_input || ev.toolInput || {};
      const st = toolState(tn, ti);
      log("  tool=" + tn + " → " + st);
      post("/api/state", { ...base, state: st });
      break;
    }
    case "post_tool": {
      const ev = await readStdin();
      if (!ev) { post("/api/state", { ...base, state: "thinking" }); break; }
      const tr = ev.tool_response || ev.toolResponse || {};
      const err = typeof tr === "object" ? (tr.isError || tr.is_error) : false;
      post("/api/state", { ...base, state: err ? "error" : "thinking" });
      break;
    }
    case "stop":
      post("/api/direct", { state: "success" });
      break;
    case "subagent_stop":
      post("/api/state", { ...base, state: "thinking" });
      break;
    case "notification":
      break;
  }
}

main();

File: adapters/claude-code/hooks.py
#!/usr/bin/env python3
"""Claude Code hook → bridge relay. Minimal imports for fast startup on Windows.

Remote usage (LAN): set env vars to point at another machine's bridge.
  YEELIGHT_RELAY_URL=http://192.168.x.x:9877
  YEELIGHT_API_KEY=<api-key>
"""
import json, sys, os, time
from urllib.request import Request, urlopen

BRIDGE = os.environ.get("YEELIGHT_RELAY_URL", "http://127.0.0.1:9877")
API_KEY = os.environ.get("YEELIGHT_API_KEY", "")
LOG = os.path.expanduser("~/.yeelight-vibe-bridge/hook_debug.log")
READ_TOOLS = {"Read","LS","Grep","Glob","Task","TodoRead","NotebookRead"}
WRITE_TOOLS = {"Write","Edit","NotebookEdit"}
EXEC_TOOLS = {"Bash","PowerShell"}
WEB_TOOLS = {"WebSearch","WebFetch"}
WEB_KW = ["curl","wget","http ","fetch","npx ","npm ","pip "]

def post(path, data):
    try:
        body = json.dumps(data).encode()
        req = Request(f"{BRIDGE}{path}", data=body, method="POST")
        req.add_header("Content-Type", "application/json")
        if API_KEY:
            req.add_header("Authorization", f"Bearer {API_KEY}")
        urlopen(req, timeout=0.3)
        with open(LOG, "a") as f:
            f.write(f"[{time.time():.0f}] {path} {data.get('state','?')} OK\n")
    except Exception as e:
        with open(LOG, "a") as f:
            f.write(f"[{time.time():.0f}] {path} {data.get('state','?')} ERROR: {e}\n")

def read_event():
    try:
        raw = sys.stdin.buffer.readline().decode("utf-8")
        return json.loads(raw) if raw.strip() else None
    except Exception: return None

def tool_state(name, inp=None):
    if name in READ_TOOLS: return "reading"
    if name in WRITE_TOOLS: return "writing"
    if name in EXEC_TOOLS:
        cmd = str(inp.get("command","") if isinstance(inp, dict) else "")
        return "fetching" if any(kw in cmd.lower() for kw in WEB_KW) else "executing"
    if name in WEB_TOOLS: return "fetching"
    return "thinking"

def user_prompt():
    post("/api/state", {"state":"thinking","pid":"claude-hook"})

def pre_tool():
    ev = read_event()
    if not ev:
        return post("/api/state", {"state":"thinking","pid":"claude-hook"})
    # Check all possible permission field names
    perm = (ev.get("permissionDecision") or ev.get("permission_decision")
         or ev.get("permission_required") or ev.get("requires_permission")
         or ev.get("permission") or ev.get("decision") or "")
    if perm == "ask":
        return post("/api/state", {"state":"waiting","pid":"claude-hook"})
    tn = ev.get("tool_name") or ev.get("toolName","")
    ti = ev.get("tool_input") or ev.get("toolInput",{})
    post("/api/state", {"state":tool_state(tn, ti),"pid":"claude-hook"})

def post_tool():
    ev = read_event()
    if not ev:
        return post("/api/state", {"state":"thinking","pid":"claude-hook"})
    tr = ev.get("tool_response") or ev.get("toolResponse",{})
    err = tr.get("isError") or tr.get("is_error",False) if isinstance(tr,dict) else False
    post("/api/state", {"state":"error" if err else "thinking","pid":"claude-hook"})

def stop():
    """Clear session then show success so it's not overwritten by stale thinking."""
    post("/api/state", {"state":"idle", "pid":"claude-hook"})
    post("/api/direct", {"state":"success"})

def subagent_stop():
    post("/api/state", {"state":"thinking","pid":"claude-hook"})

def notification():
    pass

H = {"user_prompt":user_prompt,"pre_tool":pre_tool,"post_tool":post_tool,
     "stop":stop,"subagent_stop":subagent_stop,"notification":notification}

if __name__ == "__main__":
    mode = sys.argv[1].lower() if len(sys.argv) > 1 else ""
    if mode in H: H[mode]()
